### PR TITLE
test: Check large environment

### PR DESCRIPTION
**Pull request by martinpitt**
With a bunch of `$XDG_DATA_DIRS`, long `$PATH` and `$LS_COLORS` it is    
easy to exceed 4KB in the environment. Before commit 85c805c61a13, this    
made the agent (and most of the tests) crash with    
    
    SyntaxError: unterminated string literal (detected at line 1)    
    
when reading the request.    
    
Add a unit test with a large environment variable to ensure that this    
does not regress again.

### CODE REVIEW

File: src/ferny/interaction_agent.py [hunk 1]
```diff
@@ -251,6 +251,15 @@ async def communicate(self) -> None:
             try:
                 # We handle fds very carefully to avoid leaking them, even in case of exceptions
                 data, fds, _flags, _addr = recv_fds(self.ours, 4096, 10)
+                # read the rest of the request
+                while True:
+                    try:
+                        block = self.ours.recv(4096, socket.MSG_DONTWAIT)
+                        if not data:
+                            break
```
Comment (martinpitt) [inline src/ferny/interaction_agent.py:259]: In my experiments and the unit tests, this is the normal code path, i.e. this runs into EOF. I wonder if we should actually do blocking reads until EOF -- does this *ever* expect any further communication from the client? When I do that naïvely, `test/test_basic.py::test_connection_refused` just hangs indefinitely, so it seems there are at least *some* cases where we don't get an EOF.

**Comment by martinpitt** [comment]
Hmm, `test_dns_error` is annoyingly flaky, it [tends to hang](https://github.com/allisonkarlitskaya/ferny/actions/runs/4805802065/jobs/8552739161) indefinitely..

File: src/ferny/interaction_agent.py [hunk 2]
```diff
@@ -251,6 +251,15 @@ async def communicate(self) -> None:
             try:
                 # We handle fds very carefully to avoid leaking them, even in case of exceptions
                 data, fds, _flags, _addr = recv_fds(self.ours, 4096, 10)
+                # read the rest of the request
+                while True:
+                    try:
+                        block = self.ours.recv(4096, socket.MSG_DONTWAIT)
```
Comment (allisonkarlitskaya) [inline src/ferny/interaction_agent.py:257]: I had the impression (from testing) that the fds would always arrive with the last byte of the message that they were sent with.  Maybe that's not the case?

I'd like to test this very thoroughly to make sure we know exactly what the ordering guarantee is here.  As its written, this loop could easily pick up data from the following request.

Comment (martinpitt) [reply src/ferny/interaction_agent.py:257]: A simpler approach would of course just be to increase the 4096 size of the first read to something really large, maybe 65536? But that would just make the bug less likely to happen, not really fix it.

My immediate selfish reason was "I want `pytest` to succeed", but for production quality we should support reading the request in multiple chunks IMHO.

But good point about the fds being in the last chunk! The test case seems to work (albeit with the random hang, see above), and locally `pytest` is always happy, but of couse we can add more corner cases.

Comment (allisonkarlitskaya) [reply src/ferny/interaction_agent.py:257]: We actually *do* support reading multiple chunks.  We append `data` to `self.buffer` and only once we receive the `fds` then we do the interaction based on the accumulated `self.buffer`.

Comment (martinpitt) [reply src/ferny/interaction_agent.py:257]: Then demonstrably the fds get sent with the first block, not the last one. Are you saying that `pytest` works for you with a large environment? (I'm not actually how exactly `$LS_COLORS` gets into my environment, but it should be easy enough to define a large env variable -- that's exactly what the unit test does)

Comment (allisonkarlitskaya) [reply src/ferny/interaction_agent.py:257]: Nope.  It absolutely fails for me too.

Comment (allisonkarlitskaya) [reply src/ferny/interaction_agent.py:257]: I did a test.  It seems that, more or less, the file descriptors end up associated with the first byte of the data that they were sent with.  Whenever that byte gets read off the wire, you get your fds.

It's tricky, though: the kernel will never combine lists of fds (ie: reading multiple small fd-containing messages will effectively get you packet semantics) but it's happy to combine non-fd-containing messages that came before.

Ie: if you:

  - send 'abc'
  - send '123' + fd

then the kernel will happily give you 'abc123' + fd.

This is a headache...

**Comment by martinpitt** [comment]
@allisonkarlitskaya : I dropped the code change and left the unit test. Description/commit message updated.

**Review by allisonkarlitskaya — APPROVED** [review]
Thanks! 💕
